### ISSUE TITLE
Too Easy To Destroy Cluster - High Level Discussion

### ISSUE THREAD

**Issue by benyanke** [rancher/rancher#15060]
I've run into this issue several times, and it seems to be a common thread through many issues (#15057 , #11991 ) , so I wanted to bring it up as a broader discussion to see what the best way to handle would be, or perhaps clarify if I am severely misunderstanding.

Rancher is a layer to manage one or more kubernetes cluster. However, it seems far too easy to be able to put the kubernetes cluster in an unrecoverable state.

For example, I tried the following:

1) Make a new cluster. 3 worker nodes, 3 "master" nodes (etc + control plane). 

2) From the rancher gui, delete each of the three master nodes.

3) Automatically provisioned replacement master nodes become stuck in state `Waiting for node to be removed from cluster; waiting on node-controller`. 

This is, I assume, because the masters are waiting to be successfully registered from the control plane, which is itself no longer online. The entire cluster becomes stuck and apparently unrecoverable without a complete nuke'n'pave of the cluster.

I would assume a similar situation would happen if the underlying master VMs were to go away on their own (for example, being deleted from the AWS console without being registered from rancher).
 
While this is somewhat understandable, it's also a big flaw. Is there any way to currently work around this, or am I misunderstanding something about the setup? Or is this the actual desired behavior, and you pray your masters never go down? 

Is there anything rancher can do behind the scenes to fix this transition of unreachable control plane? As it is now, it seems that if the master nodes can't reach quorum, your cluster is dead forever, as it can't rebuild itself.

**Comment by aric49** [comment]
New Rancher user/long time Kubernetes user -   I totally agree with your sentiments and observations. I was thinking the same thing myself when creating a few testing clusters. It seems to be like a pretty easy fix from a UI perspective -- simply add a warning dialog when the user attempts to delete master nodes which would cause the cluster to lose ETCD quorum, or go into a unrecoverable state. 

A longer term solution (I think) would be to take action on control-plane healthcheck failures which would cause Rancher to proactively spin up new master nodes, or re-spawn the control plane containers  in the event the control plane becomes unreachable.   Then of course, we have the issue of ETCD backups.   If the user (or Rancher) hasn't taken an ETCD backup prior to deleting the nodes, Rancher would never be able to truly restore the cluster state.   I think for now, the fastest turn around would be to add a dialog to the server UI which reads:   `Warning:  Deleting master nodes may cause your cluster to become unstable or unreachable. Proceed with caution and backup ETCD cluster state before deleting master nodes.`

**Comment by benyanke** [comment]
I'm a newish user to both Rancher and Kubernetes, and you hit the nail on the head for what I'm talking about.

Rancher is big into abstracting away the complexity of kubernetes (which it's fairly good at, to be fair), but this is a major blindspot.

I think this could be nearly-handled by the things you mention:

1) Regular ETCD backups - on the order of every minute or so
2) Error recovery from unrecoverable health check outages - this is easier on AWS, could be more difficult with self-provision nodes, but very much needed.
3) Related to (2), handling of how to add new master nodes if the control plane itself isn't online to register master nodes into.

**Comment by superseb** [comment]
Thanks for bringing this up, I thought I had opened an issue on the guidance for clusters but couldn't find it so I made it here: https://github.com/rancher/rancher/issues/15075

Regarding the controlplane issue, it sounds like https://github.com/rancher/rancher/issues/13706 but wasn't reproduced. If you can file a new issue with the steps to reproduce we can take a look at it.
